test/jdk/jdk/jfr/event/io/TestSocketAdapterEvents.java

```diff
@@ -0,0 +1,119 @@
+import java.util.List;
+
+import jdk.jfr.Recording;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.test.lib.jfr.Events;
+import jdk.test.lib.thread.TestThread;
+import jdk.test.lib.thread.XRun;
+
+/**
+ * @test
+ * @key jfr
+ * @requires vm.hasJFR
+ * @library /test/lib /test/jdk
+ * @run main/othervm jdk.jfr.event.io.TestSocketAdapterEvents
+ */
+public class TestSocketAdapterEvents {
+    private static final int writeInt = 'A';
+    private static final byte[] writeBuf = { 'B', 'C', 'D', 'E' };
+
+    private List<IOEvent> expectedEvents = new ArrayList<>();
+
+    private synchronized void addExpectedEvent(IOEvent event) {
+        expectedEvents.add(event);
+    }
+
+    public static void main(String[] args) throws Throwable {
+        new TestSocketAdapterEvents().test();
+    }
+
+    public void test() throws Throwable {
+        try (Recording recording = new Recording()) {
+            try (ServerSocketChannel ss = ServerSocketChannel.open()) {
+                recording.enable(IOEvent.EVENT_SOCKET_READ).withThreshold(Duration.ofMillis(0));
+                recording.enable(IOEvent.EVENT_SOCKET_WRITE).withThreshold(Duration.ofMillis(0));
+                recording.start();
+
+                ss.socket().setReuseAddress(true);
+                ss.socket().bind(null);
+
+                TestThread readerThread = new TestThread(new XRun() {
+                    @Override
+                    public void xrun() throws IOException {
+                        byte[] bs = new byte[4];
+                        try (SocketChannel sc = ss.accept(); Socket s = sc.socket(); InputStream is = s.getInputStream()) {
+                            int readInt = is.read();
+                            assertEquals(readInt, writeInt, "Wrong readInt");
+                            addExpectedEvent(IOEvent.createSocketReadEvent(1, s));
+
+                            int bytesRead = is.read(bs, 0, 3);
+                            assertEquals(bytesRead, 3, "Wrong bytesRead partial buffer");
+                            addExpectedEvent(IOEvent.createSocketReadEvent(bytesRead, s));
+
+                            bytesRead = is.read(bs);
+                            assertEquals(bytesRead, writeBuf.length, "Wrong bytesRead full buffer");
+                            addExpectedEvent(IOEvent.createSocketReadEvent(bytesRead, s));
+
+                            // Try to read more, but writer have closed. Should
+                            // get EOF.
+                            readInt = is.read();
+                            assertEquals(readInt, -1, "Wrong readInt at EOF");
+                            addExpectedEvent(IOEvent.createSocketReadEvent(-1, s));
+                        }
+                    }
+                });
+                readerThread.start();
+
+                try (SocketChannel sc = SocketChannel.open(ss.socket().getLocalSocketAddress()); Socket s = sc.socket(); OutputStream os = s.getOutputStream()) {
```

Comment:
If you put a line break after each ; then it would be easier to read. Also you can use ssc.getLocalAddress().